project/Component/NefJekyll/NefJekyll.swift

```diff
@@ -10,102 +10,173 @@ import Bow
+        pages.traverse { info in self.writePage(page: info.page, content: info.output, output: output)^ }^
+    }
+    
+    private func writePage(page: RenderingURL, content: RenderingOutput, output: URL) -> EnvIO<Environment, RenderError, URL> {
+        EnvIO { env in
+            let file = output.appendingPathComponent(page.escapedTitle).appendingPathComponent("README.md")
+            return env.persistence.writePage(content, file).provide(env.fileSystem)
+                                  .map { _ in file }^.mapLeft { _ in .page(file) }
+        }^
+    }
+    
+    private func writePlaygrounds(_ playgrounds: PlaygroundsOutput, into output: URL) -> EnvIO<Environment, RenderError, NEA<URL>> {
+        playgrounds.traverse { info in self.writePlayground(playground: info.playground, content: info.output, output: output)^ }^
+    }
+    
+    private func writePlayground(playground: RenderingURL, content: PlaygroundOutput, output: URL) -> EnvIO<Environment, RenderError, URL> {
+        content.traverse { info in self.writePage(page: info.page, content: info.output, output: output) }
+               .map { _ in playground.url }^
+    }
+    
+    func read(file: URL) -> EnvIO<FileSystem, RenderError, String> {
+        EnvIO { fileSystem in
+            fileSystem.readFile(atPath: file.path).mapLeft { _ in .page(file) }
+        }
+    }
+    
+    func writePage(_ page: RenderingOutput, into file: URL) -> EnvIO<Environment, RenderError, Void> {
+        EnvIO { env in
+            env.persistence.writePage(page, file).provide(env.fileSystem).mapLeft { _ in .page(file) }
+        }
+    }
+    
+    // MARK: private <renders>
+    func renderPage(content: String, permalink: String) -> EnvIO<Environment, RenderError, RenderingOutput> {
+        EnvIO { env in env.render.page(content: content).provide(env.jekyllEnvironment(permalink)) }
+    }
+    
+    func renderPlayground(_ playground: URL) -> EnvIO<Environment, RenderError, PlaygroundOutput> {
+        EnvIO { env in env.render.playground(playground).provide(env.renderEnvironment) }
+    }
+    
+    func renderPlaygrounds(atFolder folder: URL) -> EnvIO<Environment, RenderError, PlaygroundsOutput> {
+        EnvIO { env in env.render.playgrounds(at: folder).provide(env.renderEnvironment) }
+    }
+    
+    // MARK: private <steps>
+    private func buildMainPage(_ mainPage: URL, docs: URL) -> EnvIO<Environment, RenderError, Void> {
+        let file = docs.appendingPathComponent("README.md")
+        let content = (try? String(contentsOf: mainPage)) ?? """
+                                                             ---
+                                                             layout: docs
+                                                             permalink: /docs/
+                                                             ---
+                                                             """
```

Comment:
This string hanging out in the middle of the code... I think it should be extracted to a constant, or at least make sure to align it to the left of the code.